src/kcas.ml

```diff
@@ -8,181 +8,167 @@ module type Backoff = Backoff.S
-  match r with
-  | RDCSS_DESC rd ->
-      ignore @@ complete rd ;
-      rdcss_read a
-  | _ -> r
-
-let rec casn_proceed c =
-  let rec phase1 curr_c_l curr_st out =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_t_tail when curr_st = SUCCEEDED -> (
-        let s = rdcss (mk_rdcss c.st (WORD UNDECIDED) a o (CASN_DESC c)) in
-        match s with
-        | CASN_DESC c' ->
-            if c.id_casn != c'.id_casn then (
-              ignore @@ casn_proceed c' ;
-              phase1 curr_c_l curr_st out )
-            else phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-        | _ ->
-            if st_eq s o then
-              phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-            else phase1 curr_c_t_tail FAILED (CAS (a, o, n) :: out) )
-    | _ ->
-        ignore @@ commit (CAS (c.st, WORD UNDECIDED, WORD curr_st)) ;
-        out
-  in
-  let rec phase2 curr_c_l succ =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_l_tail -> (
-      match Atomic.get succ with
-      | WORD SUCCEEDED ->
-          ignore @@ commit (CAS (a, CASN_DESC c, n)) ;
-          phase2 curr_c_l_tail succ
-      | _ ->
-          ignore @@ commit (CAS (a, CASN_DESC c, o)) ;
-          phase2 curr_c_l_tail succ )
-    | _ -> Atomic.get succ = WORD SUCCEEDED
-  in
-  match Atomic.get c.st.content with
-  | WORD UNDECIDED -> phase2 (phase1 c.c_l SUCCEEDED []) c.st.content
-  | _ -> phase2 c.c_l c.st.content
-
-let rec get a =
-  let r = rdcss_read a.content in
-  match r with
-  | CASN_DESC c ->
-      ignore @@ casn_proceed c ;
-      get a
-  | WORD out -> out
-  | _ -> assert false
-
-let kCAS c_l =
-  match c_l with
+let set atom value =
+  Atomic.set atom.state { before = value; after = value; casn = casn_after }
+
+let rec release_after = function
+  | [] -> true
+  | CAS (_, cas) :: cass ->
+      cas.casn <- casn_after;
+      cas.before <- cas.after;
+      release_after cass
+
+let rec release_before = function
+  | [] -> false
+  | CAS (_, cas) :: cass ->
+      cas.casn <- casn_before;
+      cas.after <- cas.before;
+      release_before cass
+
+(* Note: The writes to `cas.casn <- ...` above could be removed to reduce
+   time at the cost of increasing space usage (by a constant factor). *)
+
+let release cass = function
+  | `After -> release_after cass
+  | `Before -> release_before cass
+
+let finish casn (`Undetermined cass as undetermined)
+    (status : [ `Before | `After ]) =
+  if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
+  then release cass status
+  else Atomic.get casn == `After
+
+let rec determine casn undetermined = function
+  | CAS (atom_state, state) :: cass_1 as cass_0 ->
+      let state' = Atomic.get atom_state in
+      if state == state' then determine casn undetermined cass_1
+      else
+        let before = state.before in
+        let before' = state'.before and after' = state'.after in
+        if
+          (before == before' && before == after')
```

Comment:
Do we lose linearization on "noop cas", i.e. one where before = after, since it looks done before the entire kCAS completes?

So for something like [cas a 0 0; cas a 0 1] and [cas a 0 1] both could succeed even if there was never a state a=0, b=1.

Maybe worth disallowing "noop" inputs?

Reply:
Amazingly, it seems to be an actual use-case: https://github.com/ocaml-multicore/lockfree/blob/a72d081e550b589cb3106f87d4b345f5e1b82470/src/paradict.ml#L132

Reply:
Note that the `before'` and `after'` values are read from the state of the atom rather than the input.

Reply:
> So for something like `[cas a 0 0; cas b 0 1]` and `[cas a 0 1]` both could succeed even if there was never a state a=0, b=1.

Hmm... Took me a moment to think about what you might mean here.

At the start of kcas a new `state` record is allocated for each ref (or individual cas) in the operation and then those states are installed to the refs.  For a kcas to succeed, all of the states need to be installed and the status updated to `After`.  Which means that it is not possible for two operations to both succeed without the intermediate state.